### PR TITLE
Add new filter 'mu_plugins'

### PR DESCRIPTION
**Pull request by nate-allen**
Adds a new filter called `mu_plugins` that allows the filtering of the must-use plugins array.

Trac ticket: https://core.trac.wordpress.org/ticket/57278

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/plugin.php [hunk 1]
```diff
@@ -407,7 +407,14 @@ function get_mu_plugins() {
 
 	uasort( $wp_plugins, '_sort_uname_callback' );
 
-	return $wp_plugins;
+	/**
+	 * Filters the array of must-use plugins.
+	 *
+	 * @since 6.2.0
+	 *
+	 * @param array $wp_plugins An array of arrays of mu-plugin data, keyed by plugin file name.
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/plugin.php]:
```suggestion
	 * @param array[] $wp_plugins An array of arrays of mu-plugin data, keyed by plugin file name.
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @nate-allen, Left one nit-pick document changes. Can you please add unit tests for the new filter?

File: tests/phpunit/tests/admin/wpPluginsListTable.php [hunk 2]
```diff
@@ -56,4 +75,31 @@ public function test_get_views_should_return_views_by_default() {
 
 		$this->assertSame( $expected, $actual );
 	}
+
+	/**
+	 * @ticket 57278
```
Comment (costdev) [inline tests/phpunit/tests/admin/wpPluginsListTable.php:107]:
```suggestion
	 * Tests that WP_Plugins_List_Table::prepare_items() 
	 * applies 'plugins_list' filters.
	 *
	 * @ticket 57278
	 *
	 * @covers WP_Plugins_List_Table::prepare_items
```

File: tests/phpunit/tests/admin/wpPluginsListTable.php [hunk 3]
```diff
@@ -56,4 +75,31 @@ public function test_get_views_should_return_views_by_default() {
...
+		remove_filter( 'plugins_list', array( $this, 'plugins_list_filter' ), 10 );
+
+		// Restore to default.
+		$status = $old_status;
+		$this->table->prepare_items();
+
+		$this->assertSame( $plugins, $this->fake_plugin );
+	}
+
+	public function plugins_list_filter( $plugins_list ) {
```
Comment (costdev) [inline tests/phpunit/tests/admin/wpPluginsListTable.php:136]:
```suggestion
	/**
	 * Adds a fake plugin to an array of plugins.
	 *
	 * Used as a callback for the 'plugins_list' hook.
	 *
	 * @return array
	 */
	public function plugins_list_filter( $plugins_list ) {
```

File: src/wp-admin/includes/class-wp-plugins-list-table.php [hunk 4]
```diff
@@ -292,11 +292,20 @@ public function prepare_items() {
 			}
 		}
 
-		if ( strlen( $s ) ) {
+		if ( isset( $s ) && strlen( $s ) ) {
```
Comment (costdev) [inline src/wp-admin/includes/class-wp-plugins-list-table.php:294]: This change doesn't appear to be related to the addition of the new filter. @nate-allen Was this part of a patch for another ticket?

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the PR @nate-allen! I've left some feedback in this review.

File: src/wp-admin/includes/class-wp-plugins-list-table.php (hunk 4 above)
Comment (nate-allen) [reply src/wp-admin/includes/class-wp-plugins-list-table.php:294]: Hi @costdev. It's required for the unit test for this ticket.

Comment (costdev) [reply src/wp-admin/includes/class-wp-plugins-list-table.php:294]: Thanks @nate-allen!

Pinging @TimothyBJacobs and @spacedmonkey as REST API maintainers. Are you happy for this source change to be made here to allow PHPUnit tests to pass, or should it be made in a separate ticket, or should the test(s) instead set a value for the `$s` global?

File: src/wp-admin/includes/class-wp-plugins-list-table.php [hunk 5]
```diff
@@ -292,11 +292,20 @@ public function prepare_items() {
...
 			$status            = 'search';
 			$plugins['search'] = array_filter( $plugins['all'], array( $this, '_search_callback' ) );
 		}
 
+		/**
+		 * Filters the array of plugins for the list table.
+		 *
+		 * @since 6.2.0
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/class-wp-plugins-list-table.php]:
```suggestion
		 * @since 6.3.0
```

Nit-pick

File: src/wp-admin/includes/class-wp-plugins-list-table.php (hunk 4 above)
Comment (TimothyBJacobs) [reply src/wp-admin/includes/class-wp-plugins-list-table.php:294]: @costdev This doesn't look REST API related to me? But since I'm here 🙂.

To me it looks like the relevant globals should be getting initialized in the test. Though I think @SergeyBiryukov or @hellofromtonya could give a more informed opinion on the direction for these kinds of unit tests.

Comment (costdev) [reply src/wp-admin/includes/class-wp-plugins-list-table.php:294]: Thanks @TimothyBJacobs! It appears my brain skipped a beat when I posted that at 3AM 😅

Yeah that's what I was thinking - and that the `isset( $s )` check, if it's going to be added at some point, should probably be separate rather than bundled in the addition of a new filter.

Let's see what Sergey or Tonya think too 🙂

Comment (hellofromtonya) [reply src/wp-admin/includes/class-wp-plugins-list-table.php:294]: Thanks for ping @TimothyBJacobs and @costdev.

>To me it looks like the relevant globals should be getting initialized in the test.

In this PR, yes, I agree. 

How to do it?

- Store each global's original value in a new property added to the test class. 
    - Add the property.
    - The initialization of the property in either the `before_set_up_class()` or `set_up()` fixture method.
- The test is then able to modify the global to its needs.
-  How to restore / reset the original value? In the `tear_down()`, set the global to its the original value using the property. 
    - Why the `tear_down()` and not within the test method? The `tear_down()` ensures the global gets reset, as it runs after each test regardless if the test fails and skips any of it test code.

[Here's an example test](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/webfonts/wpThemeJsonWebfontsHandler.php).

Comment (nate-allen) [reply src/wp-admin/includes/class-wp-plugins-list-table.php:294]: Thank you @hellofromtonya (and everyone else that has helped move this along!) I will make these changes and reach out if I have any issues.

**Comment by audrasjb** [comment]
I committed fb4a416 to solve a conflict against trunk.

**Comment by audrasjb** [comment]
closing in favor of #4717
